include/zephyr/drivers/mipi_dbi.h

```diff
@@ -0,0 +1,260 @@
+ * The current driver interface only supports type C modes 1 and 3
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_MIPI_DBI_H_
+#define ZEPHYR_INCLUDE_DRIVERS_MIPI_DBI_H_
+
+/**
+ * @brief MIPI-DBI driver APIs
+ * @defgroup mipi_dbi_interface MIPI-DBI driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/display/mipi_display.h>
+#include <zephyr/drivers/spi.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * SPI 3 wire (Type C1). Uses 9 write clocks to send a byte of data.
+ * The bit sent on the 9th clock indicates whether the byte is a
+ * command or data byte
+ *
+ *
+ *           .---.   .---.   .---.   .---.   .---.   .---.   .---.   .---.
+ *     SCK  -'   '---'   '---'   '---'   '---'   '---'   '---'   '---'   '---
+ *
+ *          -.---.---.---.---.---.---.---.---.---.---.---.---.---.---.---.
+ *     DOUT  |D/C| D7| D6| D5| D4| D3| D2| D1| D0|D/C| D7| D6| D5| D4|...|
+ *          -'---'---'---'---'---'---'---'---'---'---'---'---'---'---'---'
+ *           | Word 1                            | Word n
+ *
+ *          -.								 .--
+ *     CS    '-----------------------------------------------------------'
+ */
+#define MIPI_DBI_MODE_SPI_3WIRE 0x1
+/**
+ * SPI 4 wire (Type C3). Uses 8 write clocks to send a byte of data.
+ * an additional C/D pin will be use to indicate whether the byte is a
+ * command or data byte
+ *
+ *           .---.   .---.   .---.   .---.   .---.   .---.   .---.   .---.
+ *     SCK  -'   '---'   '---'   '---'   '---'   '---'   '---'   '---'   '---
+ *
+ *          -.---.---.---.---.---.---.---.---.---.---.---.---.---.---.---.---.
+ *     DOUT  | D7| D6| D5| D4| D3| D2| D1| D0| D7| D6| D5| D4| D3| D2| D1| D0|
+ *          -'---'---'---'---'---'---'---'---'---'---'---'---'---'---'---'---'
+ *           | Word 1                        | Word n
+ *
+ *          -.								     .--
+ *     CS    '---------------------------------------------------------------'
+ *
+ *          -.-------------------------------.-------------------------------.-
+ *     CD    |             D/C               |             D/C               |
+ *          -'-------------------------------'-------------------------------'-
+ */
+#define MIPI_DBI_MODE_SPI_4WIRE 0x2
+
+/**
+ * @brief initialize a MIPI DBI SPI configuration struct from devicetree
+ *
+ * This helper allows drivers to initialize a MIPI DBI SPI configuration
+ * structure using devicetree.
+ * @param node_id Devicetree node identifier for the MIPI DBI device whose
+ *                struct spi_config to create an initializer for
+ * @param operation_ the desired operation field in the struct spi_config
+ * @param delay_ the desired delay field in the struct spi_config's
+ *               spi_cs_control, if there is one
+ */
+#define MIPI_DBI_SPI_CONFIG_DT(node_id, operation_, delay_)		\
+	{								\
+		.frequency = DT_PROP(node_id, spi_max_frequency),	\
+		.operation = (operation_) |				\
+			DT_PROP(node_id, duplex) |			\
```

Comment:
I am unable to build our MIPI DBI driver implementation unless I am proceeding with the following changes:

<img width="394" alt="modification_1" src="https://github.com/zephyrproject-rtos/zephyr/assets/13284615/b42c1ddb-e8f0-40cd-90fa-7a4a0a859e8a">
<img width="482" alt="modification_2" src="https://github.com/zephyrproject-rtos/zephyr/assets/13284615/721055cc-d7bd-4cea-bffd-04d6fe53658d">

I am also getting some warning related to SPI bus properties for instance: "Warning (spi_bus_bridge): /soc/lcdc@30030000: node name for SPI buses should be 'spi'"

Without the changes above I am getting build errors related to SPI bus properties as indicated below:

![build_errors](https://github.com/zephyrproject-rtos/zephyr/assets/13284615/3dc509aa-0cae-4f17-8548-da0af6071bcd)

Let me know if more info is required.

Reply:
> I am also getting some warning related to SPI bus properties for instance: "Warning (spi_bus_bridge): /soc/lcdc@30030000: node name for SPI buses should be 'spi'"

Thanks for highlighting this. Internally we are building our Zephyr image with `list(APPEND EXTRA_DTC_FLAGS "-Wno-spi_bus_bridge")`. This will silence the warning you are seeing about the spi bus bridge, but I am not sure this is the best approach. Let me think about this, and see if there is a way we can avoid the devicetree issue here.


> Without the changes above I am getting build errors related to SPI bus properties as indicated below:

It's not immediately clear to me why this would be fixed by removing the `spi-device.yaml` from the devicetree. Rotation should be a property present whether or not `spi-device.yaml` is present

Reply:
> Thanks for highlighting this. Internally we are building our Zephyr image with list(APPEND EXTRA_DTC_FLAGS "-Wno-spi_bus_bridge")

Correct. Adding the mentioned DTC flag in `pre_dt_board.cmake` should do the trick. Do you know why this warning is thrown? 

> It's not immediately clear to me why this would be fixed by removing the spi-device.yaml from the devicetree

Here is a fragment of my devicetree MIPI DBI controller node (actually fragment of application overlay under `\boards`):

```
&mipi_dbi {
    reset-gpios = <&gpio0 30 GPIO_ACTIVE_LOW>;
    layer-format = <SMARTBOND_MIPI_DBI_L0_RGB565>;
    status = "okay";

    ili9340: ili9340@0 {
        compatible = "ilitek,ili9340";
        spi-max-frequency = <48000000>;
        status = "okay";
        reg = <0>;
        width = <240>;
        height = <320>;
        pixel-format = <ILI9XXX_PIXEL_FORMAT_RGB565>;
        rotation = <0>;
    };
};
```

Without removing the `spi-device.yaml` header I am getting a bunch of errors related to "...undeclared here (not in a function); did you mean..." for couple of properties such as: display-inversion, rotation, frame-format, spi-max-frequency etc. Am I missing something?

Reply:
Ah- I think I might have an idea. The `spi-device.yaml` is going to add an `on-bus: spi` property to the node. If the parent node's binding does not have a `bus: spi` property, the devicetree tooling won't generate macros for the child node. Let me take a look at this, I think we might want to switch to a `mipi-dbi-device.yaml/mipi-dbi-host.yaml` compatible structure to fix this